File: 1913-maximum-product-difference-between-two-pairs/1913-maximum-product-difference-between-two-pairs.cpp
class Solution {
public:
    int maxProductDifference(vector<int>& nums) {
        priority_queue<int> maxHeap;
        priority_queue<int, vector<int>, greater<int>> minHeap;

        for(int num:nums){
            if(maxHeap.size() < 2)     maxHeap.push(num);
            else{
                if(maxHeap.top() > num){
                    maxHeap.pop();
                    maxHeap.push(num);
                }
            }
            if(minHeap.size() < 2)  minHeap.push(num);
            else{
                if(minHeap.top() < num){
                    minHeap.pop();
                    minHeap.push(num);
                }
            }
        }
        int maxP = minHeap.top();
        minHeap.pop();
        maxP *=minHeap.top();

        int minP = maxHeap.top();
        maxHeap.pop();
        minP *= maxHeap.top();

        return maxP-minP;
    }
};

File: 1913-maximum-product-difference-between-two-pairs/1913-maximum-product-difference-between-two-pairs.py
class Solution:
    def maxProductDifference(self, nums: List[int]) -> int:
        maxHeap = []
        minHeap = []

        for num in nums:
            if len(maxHeap) < 2:
                heapq.heappush(maxHeap, -num)
            else:
                if -maxHeap[0] > num:
                    heapq.heappop(maxHeap)
                    heapq.heappush(maxHeap, -num)
            
            if len(minHeap) < 2:
                heapq.heappush(minHeap, num)
            else:
                if minHeap[0] < num:
                    heapq.heappop(minHeap)
                    heapq.heappush(minHeap, num)

        maxP = minHeap[0] * minHeap[1]

        minP = -maxHeap[0] * -maxHeap[1]

        return maxP - minP
